### PR TITLE
Associate institutes with multiple loqusdb instances

### LINKED ISSUE #1617: Show both RD and cancer loqus frequencies on the cancer page
From https://github.com/Clinical-Genomics/scout/issues/1460

### LINKED ISSUE #1830: Local observations for cancer as well (observations in cancer cases).
This might have a duplicate somewhere in the issue list. From the notes of the last user meeting (#1679). 

Fixing this requires having a loqusdb instance containing frequencies of cancer cases (ping @hassanfa again).

### PR DESCRIPTION
**Pull request by northwestwitch**
- Associate an institute to more than one loqusdb instance. Choose instances from institute settings
- On variant pages, display occurrences from several loqusdb instances
- No need to name one of the locusdb instances "default" in scout config file
-  FIx #1617 and fix #1830. 

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
</details>

**How to test**:
1. Deploy master branch on cg-vm1 and go to [a variant with observations in other cases](https://scout-stage.scilifelab.se/cust000/SWEDAC-2019-HG-05-Trio-NovaSeq/7076347d9ff7b06e8c24104ef14a2152)
1. Deploy this branch on cg-vm1 and make sure the variant above shows the same info (cases should be more since I've fixed an error in the code, see my comment in the code)
2. Modify the scout config file of cg-vm1 to have a 2 functioning instances of loqusdbapi (could be the same repeated twice as well)
3. Go to an institute's settings and select both instances on the dropdown menu. Save.
4. Go to the variant above and make sure stats for both loqusdb instances are visible

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [x] code approved by DN
- [x] tests executed by CR

### CODE REVIEW

File: scout/commands/update/institute.py [hunk 1]
```diff
@@ -29,7 +28,12 @@
     help="Specify email adress for a existing user that should be added to the institute",
 )
 @click.option("-d", "--display-name", help="Set a new display name for a insitute")
-@click.option("-l", "--loqusdb_id", help="Set a new loqusdb_id for a institute")
+@click.option(
```
Comment (northwestwitch) [inline scout/commands/update/institute.py:31]: Loqusdb instances could be set from cli. I've changed the option to be multiple so you can update the institute wity more than one instance also from here. Tested locally and works fine

File: scout/server/blueprints/variant/controllers.py [hunk 2]
```diff
@@ -422,6 +422,52 @@ def variant_rank_scores(store, case_obj, variant_obj):
     return rank_score_results
 
 
+def get_loqusdb_obs_cases(store, variant_obj, category, obs_families=[]):
```
Comment (northwestwitch) [inline scout/server/blueprints/variant/controllers.py:425]: I've extracted some code from the observations() function below to reduce its complexity. Created this new function with that code

File: scout/server/blueprints/variant/controllers.py [hunk 3]
```diff
@@ -442,81 +488,50 @@ def observations(store, loqusdb, case_obj, variant_obj):
...
     institute_id = variant_obj["institute"]
     institute_obj = store.institute(institute_id)
-    loqusdb_id = institute_obj.get("loqusdb_id") or "default"
-    if loqusdb.loqusdb_settings[loqusdb_id]["version"] is None:
-        flash("Could not connect to the preselected loqusdb instance", "danger")
-        return {
-            "total": "N/A",
-        }
 
-    chrom = variant_obj["chromosome"]
```
Comment (northwestwitch) [inline scout/server/blueprints/variant/controllers.py:456]: I've removed all this code used to create the loqusdb query from here and I've moved it to the loqusdb extension instead.

File: scout/server/blueprints/variant/controllers.py [hunk 4]
```diff
@@ -442,81 +488,50 @@ def observations(store, loqusdb, case_obj, variant_obj):
...
+
+    # Compose variant query to submit to one or more loqusdb instances
     category = variant_obj["category"]
     if category == "cancer":
         category = "snv"
     if category == "cancer_sv":
         category = "sv"
+    loqus_query = loqusdb.set_loqus_query(variant_obj, category)
 
-    composite_id = "{0}_{1}_{2}_{3}".format(chrom, pos, ref, alt)
```
Comment (northwestwitch) [inline scout/server/blueprints/variant/controllers.py:470]: This also went to the loqusdb extension

File: scout/server/blueprints/variant/templates/variant/variant_details.html [hunk 5]
```diff
@@ -145,43 +145,48 @@
...
             <th scope="col">Total nr.</th>
           </tr>
         </thead>
         <tbody>
-          <tr>
-            <td>{{ observations.observations|default('N/A') }}</td>
-            <td>{{ observations.homozygote|default('N/A') }}</td>
-            <td>{{ observations.total }}</td>
-          </tr>
+          {% for loqusid,obs in observations.items() %}
```
Comment (northwestwitch) [inline scout/server/blueprints/variant/templates/variant/variant_details.html:161]: Main branch, with only one instance:

image


This code creates something like this:

image

File: scout/server/extensions/loqus_extension.py [hunk 6]
```diff
@@ -148,6 +151,35 @@ def get_exec_loqus_version(self, loqusdb_id=None):
         LOG.debug("version: {}".format(version))
         return version
 
+    def set_loqus_query(self, variant_obj, category):
```
Comment (northwestwitch) [inline scout/server/extensions/loqus_extension.py:154]: This code was in the variant controller before, but I think this is a good place to have this function instead!

**Comment by northwestwitch** [comment]
This is ready for review now, tested on stage and seems to work fine!

File: scout/server/blueprints/variant/controllers.py [hunk 7]
```diff
@@ -422,6 +422,52 @@ def variant_rank_scores(store, case_obj, variant_obj):
...
+        category(str)
+        obs_families(list). List of all cases in loqusdb where variant occurred
+
+    Returns:
+        obs_cases(list).
+    """
+    obs_cases = []
+    user_institutes_ids = set([inst["_id"] for inst in user_institutes(store, current_user)])
+    for i, case_id in enumerate(obs_families):
+        if len(obs_cases) > 10:
```
Comment (northwestwitch) [inline scout/server/blueprints/variant/controllers.py:441]: Before this line was 
```
if i > 10:
break
```

But if you really want to collect 10 other cases with same variant observed this code was wrong, check at all the `continue` you have down the function. It's very likely that you will end up with less than 10 cases observed if you use i to break the loop! That's also the reason why you see more cases with observations when you use this branch.

**Comment by dnil** [comment]
Nice! One more meeting, then I can have a look.

File: scout/server/blueprints/institutes/templates/overview/utils.html [hunk 8]
```diff
@@ -101,9 +101,9 @@ <h5 class="header">General Institute Settings</h5>
               <!-- Admins can modify loqusdb settings -->
               <div class="col-md-4">
                 {{ form.loqusdb_id.label(class="control-label", data_bs_toggle="tooltip", data_bs_placement="top", title="LoqudDB id" ) }}<br>
-                <select class="selectpicker" name="loqusdb_id" data-style="btn-secondary">
+                <select multiple class="selectpicker" name="loqusdb_id" data-style="btn-secondary">
```
Comment (dnil) [inline scout/server/blueprints/institutes/templates/overview/utils.html:104]: 👍

Comment (dnil) [reply scout/server/blueprints/institutes/templates/overview/utils.html:104]: Is there a difference to `multiple="multiple"`?

File: tests/server/blueprints/variant/test_variant_controllers.py [hunk 9]
```diff
@@ -134,43 +155,81 @@ def test_observations_controller_non_existing(app, case_obj, loqusdb):
...
+    store.institute_collection.find_one_and_update(
+        {"_id": institute_obj["_id"]}, {"$set": {"loqusdb_id": loqus_id}}
+    )
+
+    # GIVEN a database with a case with a specific SV variant
     store.variant_collection.insert_one(sv_variant_obj)
-    loqusdb._add_variant(sv_variant_obj)
+
+    ## GIVEN patched request to the loqusdbapi instance
+    req_url = f"http://127.0.0.1:9000/svs/?chrom={sv_variant_obj['chromosome']}&end_chrom={sv_variant_obj['end_chrom']}&pos={sv_variant_obj['position']}&end={sv_variant_obj['end']}&sv_type={sv_variant_obj['sub_category'].upper()}"
```
Comment (dnil) [inline tests/server/blueprints/variant/test_variant_controllers.py:226]: Could we wrap the base url/port perhaps?

File: tests/server/conftest.py [hunk 10]
```diff
@@ -91,6 +58,14 @@ def app(real_database_name, real_variant_database, user_obj):
             MME_TOKEN=str(uuid.uuid4()),
             BEACON_URL="http://localhost:6000/apiv1.0",
             BEACON_TOKEN="DEMO",
+            LOQUSDB_SETTINGS={
+                "test": {
+                    "api_url": "http://127.0.0.1:9000",
```
Comment (dnil) [inline tests/server/conftest.py:63]: Like here as well.. 😊

**Review by dnil — APPROVED** [review]
Looks really good; just some minor things I believe. Oh, and fix that broken merge!

**Comment by dnil** [comment]
We should probably do a little bit more testing before merging, but really good functionality!

File: scout/server/blueprints/institutes/templates/overview/utils.html (hunk 8 above)
Comment (northwestwitch) [reply scout/server/blueprints/institutes/templates/overview/utils.html:104]: I didn't know one could specify multiple="multiple". [Here](https://www.w3schools.com/tags/att_select_multiple.asp) they use only "multiple"

File: scout/server/blueprints/variant/templates/variant/variant_details.html [hunk 11]
```diff
@@ -151,20 +151,44 @@
...
+                    ...
+                    {% endif%}
+                  {% endif %}
+              </td>
+            </tr>
+          {% endfor %}
         </tbody>
       </table>
     </div>
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/server/blueprints/variant/templates/variant/variant_details.html:190]: woopsie!

[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/server/blueprints/variant/templates/variant/variant_details.html]: Done now

File: tests/server/blueprints/variant/test_variant_controllers.py (hunk 9 above)
Comment (northwestwitch) [reply tests/server/blueprints/variant/test_variant_controllers.py:226]: Yes!

File: scout/server/extensions/loqus_extension.py [hunk 12]
```diff
@@ -148,6 +151,35 @@ def get_exec_loqus_version(self, loqusdb_id=None):
...
+        """Create a loqusDB query based on the variant's characteristics
+
+        Args:
+            variant_obj(scout.models.Variant)
+            category(str)
+        Returns:
+            loqus_query(dict)
+        """
+        chrom = variant_obj["chromosome"]
+        end_chrom = variant_obj.get("end_chrom", chrom)
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/server/extensions/loqus_extension.py:164]: This is the same as the code that we have in prod right now. I just moved it here. I think translocations with chrom==end_chrom are already covered as in that case, in the variant document end_chrom== chromosome, no?

**Comment by northwestwitch** [comment]
I'd like to merge this one in the next release, just to get rod of it. I'm going to test it on stage by editing the stage config file!

**Comment by northwestwitch** [comment]
Testing this branch in a combo with https://github.com/Clinical-Genomics/servers/pull/830

After installing the multiloqus config file on cg-vm1 stage:

image

Selected both:

image

Checked SNVs. Example: https://scout-stage.scilifelab.se/cust000/F0006358-mip7/d99a3919e08863f39ef8123db748785e

image


And SVs: (example: https://scout-stage.scilifelab.se/cust000/F0006358-mip7/sv/variants/d8611a2441f570cccd5903d05f8371c8)

image


image

File: scout/server/blueprints/institutes/templates/overview/utils.html (hunk 8 above)
Comment (northwestwitch) [reply scout/server/blueprints/institutes/templates/overview/utils.html:104]: The important thing is that they all work! :)

**Pull request merged**